### PR TITLE
Preventing illegal method calls

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Fixes: #9 
Fixes: #12 
Fixes: #6 

Exploring the possibility of throwing a clearer exception explaining what the user's mistake is.

Using the Query object incorrectly can result in an error that may not clearly reflect what the incorrect action was.

The following code:
```
var expression = new RootQuery()
    .Simple("foo")
    .Where(simple => simple.Name == "Something")
    .Select(x => x.Name);
```

Throws the following exception:

> System.NotSupportedException: Don't know how to cast 'simple => (simple.get_Item("name").ToObject() == "Something")' to 'System.Linq.Expressions.Expression`1[System.Func`2[Octokit.GraphQL.Core.UnitTests.Models.Simple,System.Boolean]]'

### CODE REVIEW

File: Octokit.GraphQL.Core/Builders/QueryBuilder.cs [hunk 1]
```diff
@@ -225,9 +225,18 @@ private Expression VisitMethodCall(MethodCallExpression node, MemberInfo alias)
             }
             else
             {
-                return node.Update(
-                    Visit(node.Object),
-                    VisitMethodArguments(node.Method, node.Arguments));
+                if (node.Method.Name == "Where")
+                {
+                    throw new NotSupportedException();
```
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core/Builders/QueryBuilder.cs]: We could catch the dis-allowed method calls and throw a better error here?... but it feels kinda dirty.

Comment (grokys) [reply Octokit.GraphQL.Core/Builders/QueryBuilder.cs]: We could just throw a `NotSupportedException` with the message "Where() not currently supported" or something?

File: Octokit.GraphQL.Core/Builders/QueryBuilder.cs [hunk 2]
```diff
@@ -231,9 +231,23 @@ private Expression VisitMethodCall(MethodCallExpression node, MemberInfo alias)
...
+
+                if (node.Method.Name == "GroupBy")
+                {
+                    throw new NotSupportedException("GroupBy() is not supported");
+                }
+
+                if (node.Method.Name == "SkipWhile")
+                {
+                    throw new NotSupportedException("SkipWhile() is not supported");
+                }
```
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core/Builders/QueryBuilder.cs]: This list can go on and on... Is there something better we can do here?

File: Octokit.GraphQL.Core/QueryProvider.cs [hunk 3]
```diff
@@ -23,7 +23,7 @@ public object Execute(Expression expression)
 
         public TResult Execute<TResult>(Expression expression)
         {
-            throw new NotImplementedException();
+            throw new NotSupportedException("QueryProvider cannot be executed");
```
Comment (StanleyGoldman) [inline Octokit.GraphQL.Core/QueryProvider.cs]: I found another weird patch that we also complained about in #6

File: Octokit.GraphQL.Core/Builders/QueryBuilder.cs (hunk 2 above)
Comment (grokys) [reply Octokit.GraphQL.Core/Builders/QueryBuilder.cs]: I'm a bit rusty on all this code, but would whitelisting the methods we _do_ accept work?

**Pull request merged**